x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/UserRoleMapper.java

```diff
@@ -222,29 +276,28 @@ public boolean test(FieldExpression.FieldValue fieldValue) {
                     final String suffix = pattern.substring(2);
                     // if the suffix has a wildcard, then it's not a pure sub-tree match
                     if (suffix.indexOf('*') == -1) {
-                        final DN dnSuffix = parseDn(suffix);
-                        if (dnSuffix != null && dn.isDescendantOf(dnSuffix, false)) {
+                        final String normalizedDnSuffix = dnNormalizer.normalize(suffix);
+                        if (normalizedDnSuffix != null && normalizedDn.endsWith("," + normalizedDnSuffix)) {
```

Comment:
I'm not 100% convinced that this is exactly the same behaviour. It might be, but I can't prove it to myself, and that worries me.

The question is, _is there a string `S`, which can be parsed as a DN, to produce a normalized string `N`, and another normalised-dn `G` for which `G` ends with `",$N"` , but `G` is not a descendant of `N`_?

I can't come up with a scenario under which that could be an issue, but I'm worried there is one.

If we want to be _absolutely safe_, then there are 2 options we could take:
1. Have this `DistinguishedNamePredicate` retain the `DN` for its value, and then walk its ancestors and see if any of them equal the `normalizedDNSuffix`. That's cumbersome, but not too intensive from a CPU or memory point of view.
2. Call  `DN.isDescendantOf(String,String,boolean) and take the re-parsing hit. That's going to be slower, but this code isn't used much so it might be justified.

Or, we can decide that the code as-is is safe enough, but I'm not 100% sure if it is.

Reply:
I found there is a case that this is not equivalent. If `normalizedDnSuffix` is a root DN, i.e. `""`, the `normalizedDnSuffix` should be descendant of it (as long as the `normalizedDnSuffix` itself is *not* a root DN). But this string comparison with `endsWith` will give opposite result.

It should be relatively easy to handle the root DN situation. Other than this, based on reading the code of `DN` and `RDN`, I don't see other problems by string comparison because `isDescendantOf` essentially checks 
```
while (thatPos >= 0)
    {
      if (! rdns[thisPos--].equals(dn.rdns[thatPos--]))
      {
        return false;
      }
    }
``` 

and `RDN.equals` is a normalised string comparision:
```
    final RDN rdn = (RDN) o;
    return (toNormalizedString().equals(rdn.toNormalizedString()));
```

Also, `DN.toNormalizedString` is just a concatenation of `RDN.toNormalizedString`:
```
for (int i=0; i < rdns.length; i++)
    {
      if (i > 0)
      {
        buffer.append(',');
      }

      buffer.append(rdns[i].toNormalizedString());
    }
```

So all these look safe to me.

Reply:
I spent some time on trying to come up with an example (other than empty root DN) that breaks the string comparison. But I could not find any. My conclulsion so far is that it is safe. 

To make the string comparision break, we need:
1. Both DN and suffixDN are well formed DN (otherwise they will fail earlier on DN parsing before reaching to string comparision)
2. The DN normalized string must ends with comma plus suffixDN normalized string, i.e. `DN = {prefix},{suffixDN}`
3. When parsing the DN (`{prefix},{suffixDN}`), the `{suffixDN}` part is somehow **not** parsed as separate RNDs from the `{prefix},` part. 

Step 3 is where most effort is needed to prove the correctness. But step 1 actually provides a lot of help because things are easier to reason with when DNs are well formed.

Because suffixDN is well formed, it must contain at least one unescaped `=` (we don't consider empty root DN here since it will be handled separately) that is between attributeType and attributeValue. This unescaped `=` must also be parsed as separator between attributeType and attributeValue for DN because DN is also well formed (there is no other way for unescaped `=` to be in a DN without failing the parsing). So for the DN, the left side of `=` must also be parsed as attributeType. 

At this point, for DN to *not* be a descendant of suffixDN, the only possibility is that DN parses the attributeType (immediately before `=`) differently from how suffixDN parses it. Since `endsWith` test not just with the attributeType but with `,{attributeType}=...` (note the leading comma). This means DN's different attributeType must be something like `...,{attributeType}`. This is impossible because attributeType cannot contain `,` (not even an escaped `,`). Therefore it is impossible for DN to parse a different attributeType since we know it is well formed. It must also parse the components after the comma (`,`) the same way as suffixDN does. This means DN must be a descendant of suffixDN. 

I tried to consider how unicode can change things. But I don't think they do. Therefore I think it's safe to go with the string comparison. What do you think?

Reply:
I also cannot find an example that breaks it. I think if one does exist it would be an extreme edge case and unlikely to ever be encountered in real life.